extensions/standard-processors/utils/JoltUtils.cpp

```diff
@@ -0,0 +1,1134 @@
+    }
+    if (is_template) {
+      if (auto templ = Spec::Template::parse(name.begin(), name.end())) {
+        if (templ->second != name.end()) {
+          throw Exception(GENERAL_EXCEPTION, fmt::format("Failed to parse template at {}, unexpected char at {}", ctx.path(), std::distance(name.begin(), templ->second)));
+        }
+        // dry eval so we can check if the references refer to valid substrings
+        (void)templ->first.eval(ctx);
+        Spec::Context sub_ctx = ctx.extend({name}, nullptr);
+        result->templates.insert({templ->first, parseValue(sub_ctx, member)});
+      } else {
+        throw Exception(GENERAL_EXCEPTION, fmt::format("Error while parsing key template at {}: {}", ctx.path(), templ.error()));
+      }
+    } else if (is_regex) {
+      if (auto reg = Spec::Regex::parse(name)) {
+        Spec::Context sub_ctx = ctx.extend({name}, nullptr);
+        sub_ctx.matches.resize(reg.value().size());
+        result->regexes.insert({reg.value(), parseValue(sub_ctx, member)});
+      } else {
+        throw Exception(GENERAL_EXCEPTION, fmt::format("Error while parsing key regex at {}: {}", ctx.path(), reg.error()));
+      }
+    } else {
+      Spec::Context sub_ctx = ctx.extend({name}, nullptr);
+      std::optional<size_t> numeric_value;
+      auto literal_name = parseLiteral(name);
+      result->literal_indices.insert({literal_name, result->literals.size()});
+      if (isAllDigits(literal_name.begin(), literal_name.end())) {
+        numeric_value = std::stoull(literal_name);
+      }
+      result->literals.push_back({literal_name, numeric_value, parseValue(sub_ctx, member)});
+    }
+  }
+}
+
+std::unique_ptr<Spec::Pattern> parseMap(const Spec::Context& ctx, const rapidjson::Value& val) {
+  if (!val.IsObject()) {
+    throw Exception(GENERAL_EXCEPTION, fmt::format("Expected a map at '{}'", ctx.path()));
+  }
+  auto map = std::make_unique<Spec::Pattern>();
+
+  enum class State {
+    Plain,
+    Escaped
+  } state = State::Plain;
+
+  for (auto& [name_val, member] : val.GetObject()) {
+    std::string_view name{name_val.GetString(), name_val.GetStringLength()};
+    std::string subkey;
+    for (size_t idx = 0; idx <= name.size(); ++idx) {
+      std::optional<char> ch;
+      if (idx < name.size()) {
+        ch = name[idx];
+      }
+      switch (state) {
+        case State::Plain: {
+          if (ch == '\\') {
+            state = State::Escaped;
+          } else if (!ch || ch == '|') {
+            parseMember(ctx, map, subkey, member);
+            subkey.clear();
+          } else {
+            subkey += ch.value();
+          }
+          break;
+        }
+        case State::Escaped: {
+          if (!ch) {
+            throw Exception(GENERAL_EXCEPTION, "Unterminated escape sequence");
+          }
+          if (ch == '|') {
+            subkey += "|";
+          } else {
+            subkey += "\\";
+            subkey += ch.value();
+          }
```

Comment:
there is a subsequent processing step that checks for escapes again, so you can put `abc\\|def\|gh` which will be divided into `['abc\\', 'def|gh']` which then processed further into `['abc\', 'def|gh']`, note that this does not work in the original implementation, you cannot escape `|` characters, they are always treated as multiple subpatterns, but since no test verified that behavior I added it here
(added some comments on the intentionality of leaving `\\`)